### ISSUE TITLE
Are BERT word-embeddings capable of synonyms?

### ISSUE THREAD

**Issue by datistiquo** [google-research/bert#507]
I am new to BERT. The main part of BERT is beeing capable of different contextual meaning for a word. But in my case I more need to be able to capture synonyms. So I ask myself if BERT better suited for accounting synonyms than like FastText? Problem with word-embedings like with fasttext is that there can be high similairity of two words which are no synonyms though.

At the level of word embedings do have there also many embeddings for same word? 

I just want to train word embedings and save them in a format loke Glove to work with my own models further beccause BERT is of course computational intensive which I cannot solve for my use case.

Thanks!

**Comment by apogiatzis** [comment]
One thing to realise is that Glove provides context-free embeddings (static) whereas BERT gives contextualised embeddings (dynamic). For instance, take these two sentences "I like apples", "I like apple MacBooks". Glove will give the same embedding for the word apple in both sentences whereas BERT will give you a different one depending on the context.

Now coming back to your question, if you save the exported embeddings then the contextualisation is gone, which in a way beats the whole purpose of BERT. But in any case, here is a step by step tutorial I wrote to obtain contextualised embeddings from BERT:
https://towardsdatascience.com/nlp-extract-contextualized-word-embeddings-from-bert-keras-tf-67ef29f60a7b

**Comment by datistiquo** [comment]
@apogiatzis Thanks. But I am in a strict specific domain where synonyms are important to tackle. There entities also occur in different contexts but the entity still is the SAME like credit card - you want order one or have lost one (2 contexts but same entity). I wonder if BERT emebdinngs are better to capture then the synonyms (if finetuned on text where all those synonyms appear).

**Comment by maneeshaindrachapa** [comment]
@datistiquo was you able to found anything that can find synonyms using BERT?

**Comment by zzks** [comment]
@datistiquo what embedding do you use to find synonyms? w2v, fasttext, glove.... which one is the best embedding for synonyms detection? and which distance is best for synonyms ranking? thanks!
I am using w2v and cosine distance.

**Comment by ldmtwo** [comment]
If I may add on, I also need a way to deal with synonyms/substitutions. I don't know how Google search does it. I realize synonyms are tricky in some cases, but I would like to naively handle the simple cases like "two"="2" and "condemn"=denounce/disapprove. I assume that fasttext would learn some synonyms from training, but only those that were presented in a very similar way in every example. I have a small dataset of synonyms, but the questions is how to make the embeddings guarantee a very small distance and how to make the training function learn these hardcoded relations in addition to unsupervised data. 

One idea I have is to duplicate the training data with random substitutions for close/exact synonyms. It's expensive and may create some garbage occasionally, but I think it's doable. It might also help the model know that 2=two, dos, deux, ....etc for each language. Maybe that would be an easier way to teach it Spanglish, Chinglish, etc?

**Comment by maayanorner** [comment]
Hi,
I am not an expert for BERT, but maybe I can help a little (I hope I will be reasonably accurate):

@ldmtwo The thing is, that I am not sure where and how you can "put the line" of handling only simple cases... It will be simple only as long as you do not have multiple relevant/frequent senses in my opinion, and I am not sure if it's possible/takes reasonable amounts of effort to tell (which words are problematic).

@zzks 
I think BERT can be a great tool if you want to find which words fit well to a given context (replacing socket with pipe will be almost reasonable if you are talking about servers, but you do not want a socket in your pool).

The thing is, w2v, fasttext, and other "traditional" embeddings, will merge multiple senses of a given word to one static vector, while in BERT they use contextualized word representations.

If you have no context words, you should use the first ones, but if you want to see which words fit **the context** well (for example: "he went to the {mask} to buy food"), you can use BERT for fill-in-the-blank task, so you can prevent incorrect usage of words (for example, "put the socket in the pull").

**Comment by maayanorner** [comment]
Hi,
I am not an expert of BERT, but maybe I can help a little (I hope I will be reasonably accurate):

@ldmtwo The thing is, that I am not sure where and how you can "put the line" of handling only simple cases... It will be simple only as long as you do not have multiple relevant/frequent senses in my opinion, and I am not sure if it's possible/takes reasonable amounts of effort to tell (which words are problematic).

@zzks
I think BERT can be a great tool if you want to find which words fit well to a given context (replacing socket with pipe will be almost reasonable if you are talking about servers, but you do not want a socket in your pool).

The thing is, w2v, fasttext, and other "traditional" embeddings, will merge multiple senses of a given word to one static vector, while in BERT they use contextualized word representations.

If you have no context words, you should use the first ones, but if you want to see which words fit the context well (for example: "he went to the {mask} to buy food" will yield "he went to the supermarket to buy food" with a high probability related to it), you can use BERT for fill-in-the-blank task, so you can prevent incorrect usage of words (for example, "put the socket in the pool").

**Comment by maayanorner** [comment]
Hi,
I am not an expert of BERT, but maybe I can help a little (I hope I will be reasonably accurate):

@ldmtwo The thing is, that I am not sure where and how you can "put the line" of handling only simple cases... It will be simple only as long as you do not have multiple relevant/frequent senses in my opinion, and I am not sure if it's possible/takes reasonable amounts of effort to tell (which words are problematic).

I think BERT can be a great tool if you want to find which words fit well in a given context (replacing socket with pipe will be almost reasonable if you are talking about servers, but you do not want a socket in your pool).

The thing is, w2v, fasttext, and other "traditional" embeddings, will merge multiple senses of a given word to one static vector, while in BERT they use contextualized word representations.

If you have no context words, you should use the first ones, but if you want to see which words fit the context well (for example: "he went to the {mask} to buy food" will yield "he went to the supermarket to buy food" with a high probability related to it), you can (though I am not sure if should) use BERT for fill-in-the-blank task or any other better way you can figure, so you can prevent incorrect usage of words (for example, "put the socket in the pool").

A simple algorithm for improved synonyms data augmentation  (not necessarily an efficient one) would be to retrieve synonyms candidates from traditional embeddings and then see which fits the context well.

> "One idea I have is to duplicate the training data with random substitutions for close/exact synonyms. It's expensive and may create some garbage occasionally"

well - if you can estimate the probability of the sentence you have just created (with synonyms), I do not see why it will not work well enough (anyways I do not think you should use BERT to estimate this probability, but correct me if I am wrong).
Or, you can just take a sentence and put a mask in it, and then get sentences with possible synonyms using fill-in-the-blank utility, then BERT is a good fit.

**Comment by meiyor** [comment]
Thank you very much to open this discussion, I'm still wondering if BERT is pre-trained without using a domain-specific text corpora the BERT capability to weight contextually synonyms or homonyms will be the same. How can we measure the incidence of the text corpora if this is more generic or not?  Exist or have you created a measure to know how reliable is BERT in generic or a domain-specific kind of pre-training environment?
